Sources/LeafKit/LeafToken.swift

```diff
@@ -181,3 +213,100 @@ public enum LeafToken: CustomStringConvertible, Equatable  {
+        precedenceMap = [
+            (check: { $0 == .not } , binary: false), // unaryNot
+            (check: { $0 == .multiply || $0 == .divide } , binary: true), // Mult/Div
+            (check: { $0 == .plus || $0 == .minus } , binary: true), // Plus/Minus
+            (check: { $0 == .greaterThan || $0 == .greaterThanOrEquals } , binary: true), // >, >=
+            (check: { $0 == .lessThan || $0 == .lessThanOrEquals } , binary: true), // <, <=
+            (check: { $0 == .equals || $0 == .notEquals } , binary: true), // !, !=
+            (check: { $0 == .and || $0 == .or } , binary: true), // &&, ||
+        ]
+            
+        groupOps: for map in precedenceMap {
+            while let i = findLastOpWhere(map.check) {
+                if map.binary { wrapBinaryOp(i)}
+                else { wrapUnaryNot(i) }
+                // Some expression could not be wrapped - probably malformed syntax
+                if ops == operandCount() { return false } else { ops -= 1 }
+                if operandCount() == 0 { break groupOps }
+            }
+        }
+        
+        flatten()
+        return ops > 1 ? true : false
+    }
+    
+    mutating func flatten() {
+        while count == 1 {
+            if case .expression(let e) = self.first! {
+                self.removeAll()
+                self.append(contentsOf: e)
+            } else { return }
+        }
+        return
+    }
+    
+    fileprivate mutating func wrapUnaryNot(_ i: Int) {
+        let rhs = remove(at: i + 1)
+        if case .parameter(let p) = rhs, case .keyword(let key) = p, key.isBooleanValued {
+            self[i] = .parameter(.keyword(Keyword(rawValue: String(!key.booleanValue!))!))
+        } else {
+            self[i] = .expression([self[i],rhs])
+        }
+    }
+    
+    // could be smarter and check param types beyond verifying non-op but we're lazy here
+    fileprivate mutating func wrapBinaryOp(_ i: Int) {
+        // can't wrap unless there's a lhs and rhs
+        guard self.indices.contains(i-1),self.indices.contains(i+1) else { return }
+        let lhs = self[i-1]
+        let rhs = self[i+1]
+        // can't wrap if lhs or rhs is an operator
+        if case .parameter(let p) = lhs, case .operator = p { return }
+        if case .parameter(let p) = rhs, case .operator = p { return }
+        self[i] = .expression([lhs, self[i], rhs])
+        self.remove(at:i+1)
+        self.remove(at:i-1)       
+    }
+
+    // Helper functions
+    func operandCount() -> Int { return reduceOpWhere { _ in true } }
+    func unaryOps() -> Int { return reduceOpWhere { $0.isUnaryPrefix } }
+    func binaryOps() -> Int { return reduceOpWhere { $0.isBinary } }
+    func reduceOpWhere(_ check: (Operator) -> Bool) -> Int {
+        return self.reduce(0, { count, pD  in
+            if let op = pD.operator() {
+                return check(op) ? count + 1 : count
+            }
+            return count
```

Comment:
Simplify as `return count + op.map { check($0) ? 1 : 0 } ?? 0`

Reply:
Done, though the map needs to be paren wrapped or Xcode chokes :)